### PR TITLE
Add self_attention.py to Implement SelfAttentionLayer for Self-Attention Architecture

### PR DESCRIPTION
**Pull request by aminejebbar**
I saw a comment from @andimarafioti  in [[this GitHub issue](https://github.com/huggingface/nanoVLM/issues/83)](https://github.com/huggingface/nanoVLM/issues/83) where they mentioned wanting to add a self-attention architecture to the project. Based on that, I created a new file `self_attention.py` and added a class `SelfAttentionLayer` that implements a configurable self-attention mechanism.

In the class, I included the option to choose between **Pre-LayerNorm** and **Post-LayerNorm** styles. The layer also supports dropout for regularization and can handle attention masks for padding tokens.

### **Changes**:

* Added `self_attention.py` with the `SelfAttentionLayer` class.
* The layer is flexible: it can use **Pre-LayerNorm** or **Post-LayerNorm**, depending on the `pre_norm` parameter.
* Attention masking support is included to handle padding tokens (e.g., in sequences).
* Dropout is also supported to help prevent overfitting.

### **Example Code**:

```python
from models.self_attention import SelfAttentionLayer

# Pre-Norm self-attention
attention_layer = SelfAttentionLayer(embed_dim=768, num_heads=12, pre_norm=True)

# Post-Norm self-attention
attention_layer_post = SelfAttentionLayer(embed_dim=768, num_heads=12, pre_norm=False)
```

### **Why This Change**:

I created this addition to help improve and experiment with the self-attention architecture. The goal is to provide a more modular and flexible self-attention layer that can easily be integrated into existing models. The `pre_norm` flag allows us to test both common normalization styles (Pre-LayerNorm vs. Post-LayerNorm) and make the model more adaptable.

### **Next Steps**:

* Please review the code in `self_attention.py`. Since I’m still learning about machine learning, any feedback on errors or potential improvements would be greatly appreciated! I’m contributing to open-source to learn and improve my skills in this area.

Also, I used ChatGPT to help write this PR so that it’s easier to read and more clear.

### CODE REVIEW

**Comment by andimarafioti** [comment]
Our current model in main is a typical `self-attention` VLM . When I said "we wanted to implement a self-attention architecture", I was referring to before we created the project. It is a self-attention architecture now.
